sdk/src/azure/core/az_span.c

```diff
@@ -983,6 +983,33 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
   }
 }
 
+AZ_NODISCARD int32_t _az_span_url_encode_calc_length(az_span source)
+{
+  _az_PRECONDITION_VALID_SPAN(source, 0, true);
+
+  int32_t const source_size = az_span_size(source);
+  if (source_size == 0)
+  {
+    return 0;
+  }
+
+  uint8_t* const src_ptr = az_span_ptr(source);
+  int32_t required_symbols_to_be_added = 0;
+  int32_t src_idx = 0;
+  do
+  {
+    uint8_t c = src_ptr[src_idx];
+    if (_az_span_url_should_encode(c))
+    {
+      // Adding '%' plus 2 digits (minus 1 as original symbol is counted as 1)
+      required_symbols_to_be_added += 2; 
+    }
+    ++src_idx;
+  } while (src_idx < source_size);
+
+  return source_size + required_symbols_to_be_added;
```

Comment:
I just realized this and ` required_symbols_to_be_added += 2;` have potential integer overflow problem for really large inputs.

**Some ways to solve this:**
1) Add a precondition that the input must be of size < int.max/3 - I think this is best given the circumstance (add a comment too)
2) Add integer overflow checks and fail with some `az_result`
3) Add integer overflow checks and fail with -1

Reply:
This is one of the things we should not pay too much attention, IMO.
int32 max is 2,147,483,647.   divided by 3 is `715,827,882`.   
We would need an url length bigger than that value to hit this edge case.

Just think about it.  If we consider the average word count in a book is 700, 000 words, and suppose an average of 10 chars per word, we would have an entire book with 7, 000, 000 chars right?  Well, that would be only the 1% of the MAX supported url.

Reply:
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Reply:
I agree with your sentiment.

That said we should be resilient to developer bugs, and edge cases with untrusted input. Adding a precondition conveys intent clearly and helps catch bugs where the user accidentally passes in corrupted or too large span (let's say they made a mistake with ptrs and size calculations (or overflowed it), even if the buffer itself isn't 800MB).

This is a very cheap check and I see no harm in adding it.

What is the behavior of this function when size is > 800 MB? Right now, undefined and error prone.
Add a precondition: Now the caller knows that size > 800 MB isn't supported.

Reply:
precondition is cheap.
Testcase is not.  I would need to allocate all that memory.  It is not required

Reply:
I would be OK with not having a test case for it. Generally, we have a lower test coverage bar when it comes to preconditions, since they themselves are a sort of check. Covering those is nice to have.

One way you can make the test (if you really wanted to), is create a busted az_span with a "fake" large size and pass that in.

Reply:
Adding a precondition would be good